File: sources/Android/android-communications/library/src/test/java/com/polar/sdk/api/model/utils/PolarTestUtilsTest.kt
package com.polar.sdk.api.model.utils

import com.polar.androidcommunications.api.ble.model.gatt.client.psftp.BlePsFtpClient
import com.polar.sdk.api.model.DeviationFromBaseline
import com.polar.sdk.api.model.Spo2Class
import com.polar.sdk.api.model.Spo2TestStatus
import com.polar.sdk.impl.utils.PolarTestUtils
import com.polar.sdk.impl.utils.Spo2TestEntry
import com.polar.services.datamodels.protobuf.Spo2TestResult.PbDeviationFromBaseline
import com.polar.services.datamodels.protobuf.Spo2TestResult.PbSpo2Class
import com.polar.services.datamodels.protobuf.Spo2TestResult.PbSpo2TestResult
import com.polar.services.datamodels.protobuf.Spo2TestResult.PbSpo2TestStatus
import io.mockk.coEvery
import io.mockk.coVerify
import io.mockk.mockk
import kotlinx.coroutines.test.runTest
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNotNull
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test
import protocol.PftpRequest
import protocol.PftpResponse.PbPFtpDirectory
import protocol.PftpResponse.PbPFtpEntry
import java.io.ByteArrayOutputStream
import java.time.LocalDate
import java.time.format.DateTimeFormatter

class PolarTestUtilsTest {

    private val dateFormatter = DateTimeFormatter.ofPattern("yyyyMMdd")

    @Test
    fun `readSpo2TestProtoFromDayDirectory() returns one entry per time subdirectory`() = runTest {
        val client = mockk<BlePsFtpClient>()
        val date = LocalDate.of(2026, 4, 8)
        val dateStr = date.format(dateFormatter)
        val spo2Dir = "/U/0/$dateStr/SPO2TEST/"

        val dirContent = buildDirectory("083906/", "112658/")
        val file1 = buildProto { setBloodOxygenPercent(97).setAverageHeartRateBpm(63).setTimeZoneOffset(180) }
        val file2 = buildProto { setBloodOxygenPercent(97).setAverageHeartRateBpm(73).setTimeZoneOffset(180) }

        coEvery { client.request(requestFor(spo2Dir)) } returns dirContent
        coEvery { client.request(requestFor("${spo2Dir}083906/SPO2TRES.BPB")) } returns file1
        coEvery { client.request(requestFor("${spo2Dir}112658/SPO2TRES.BPB")) } returns file2

        val result = PolarTestUtils.readSpo2TestProtoFromDayDirectory(client, date)

        assertEquals(2, result.size)
        assertEquals(date, result[0].date)
        assertEquals("083906", result[0].timeDirName)
        assertEquals(date, result[1].date)
        assertEquals("112658", result[1].timeDirName)

        val proto1 = PbSpo2TestResult.parseFrom(result[0].protoBytes)
        assertEquals(63, proto1.averageHeartRateBpm)
        val proto2 = PbSpo2TestResult.parseFrom(result[1].protoBytes)
        assertEquals(73, proto2.averageHeartRateBpm)
    }

    @Test
    fun `readSpo2TestProtoFromDayDirectory() returns empty list when directory listing fails`() = runTest {
        val client = mockk<BlePsFtpClient>()
        val date = LocalDate.of(2026, 4, 8)

        coEvery { client.request(any()) } throws Exception("network error")

        val result = PolarTestUtils.readSpo2TestProtoFromDayDirectory(client, date)

        assertTrue(result.isEmpty())
    }

    @Test
    fun `readSpo2TestProtoFromDayDirectory() returns empty list when SPO2TEST dir has no subdirectories`() = runTest {
        val client = mockk<BlePsFtpClient>()
        val date = LocalDate.of(2026, 4, 8)
        val dateStr = date.format(dateFormatter)
        val spo2Dir = "/U/0/$dateStr/SPO2TEST/"

        // Directory listing contains only a file, no subdirectory entries
        val emptyDir = buildDirectory()
        coEvery { client.request(requestFor(spo2Dir)) } returns emptyDir

        val result = PolarTestUtils.readSpo2TestProtoFromDayDirectory(client, date)

        assertTrue(result.isEmpty())
        coVerify(exactly = 1) { client.request(requestFor(spo2Dir)) }
    }

    @Test
    fun `readSpo2TestProtoFromDayDirectory() skips subdirectory when proto file fetch fails`() = runTest {
        val client = mockk<BlePsFtpClient>()
        val date = LocalDate.of(2026, 4, 8)
        val dateStr = date.format(dateFormatter)
        val spo2Dir = "/U/0/$dateStr/SPO2TEST/"

        val dirContent = buildDirectory("083906/", "112658/")
        val file2 = buildProto { setBloodOxygenPercent(95).setAverageHeartRateBpm(66).setTimeZoneOffset(180) }

        coEvery { client.request(requestFor(spo2Dir)) } returns dirContent
        coEvery { client.request(requestFor("${spo2Dir}083906/SPO2TRES.BPB")) } throws Exception("file not found")
        coEvery { client.request(requestFor("${spo2Dir}112658/SPO2TRES.BPB")) } returns file2

        val result = PolarTestUtils.readSpo2TestProtoFromDayDirectory(client, date)

        // Only the second subdirectory succeeds
        assertEquals(1, result.size)
        assertEquals("112658", result[0].timeDirName)
    }

    @Test
    fun `readSpo2TestProtoFromDayDirectory() returns single entry when only one time subdirectory exists`() = runTest {
        val client = mockk<BlePsFtpClient>()
        val date = LocalDate.of(2026, 4, 14)
        val dateStr = date.format(dateFormatter)
        val spo2Dir = "/U/0/$dateStr/SPO2TEST/"

        val dirContent = buildDirectory("063635/")
        val fileContent = buildProto { setBloodOxygenPercent(95).setAverageHeartRateBpm(66).setTimeZoneOffset(180) }

        coEvery { client.request(requestFor(spo2Dir)) } returns dirContent
        coEvery { client.request(requestFor("${spo2Dir}063635/SPO2TRES.BPB")) } returns fileContent

        val result = PolarTestUtils.readSpo2TestProtoFromDayDirectory(client, date)

        assertEquals(1, result.size)
        assertEquals("063635", result[0].timeDirName)
        assertEquals(date, result[0].date)
    }

    @Test
    fun `dateTimeFromFolderNames() returns correctly formatted string for valid HHMMSS`() {
        val date = LocalDate.of(2026, 4, 8)
        val result = PolarTestUtils.dateTimeFromFolderNames(date, "083906")

        assertEquals("2026-04-08 08:39:06", result)
    }

    @Test
    fun `dateTimeFromFolderNames() returns null when timeDirName is not 6 characters`() {
        val date = LocalDate.of(2026, 4, 8)
        assertNull(PolarTestUtils.dateTimeFromFolderNames(date, "0839"))
        assertNull(PolarTestUtils.dateTimeFromFolderNames(date, "08390600"))
        assertNull(PolarTestUtils.dateTimeFromFolderNames(date, ""))
    }

    @Test
    fun `dateTimeFromFolderNames() returns null when timeDirName contains non-numeric characters`() {
        val date = LocalDate.of(2026, 4, 8)
        assertNull(PolarTestUtils.dateTimeFromFolderNames(date, "AB3906"))
        assertNull(PolarTestUtils.dateTimeFromFolderNames(date, "0839XY"))
    }

    @Test
    fun `dateTimeFromFolderNames() pads single-digit hours minutes and seconds correctly`() {
        val date = LocalDate.of(2026, 4, 8)
        val result = PolarTestUtils.dateTimeFromFolderNames(date, "010203")
        assertEquals("2026-04-08 01:02:03", result)
    }

    @Test
    fun `mapSpo2TestProto() returns null testTime when folder is invalid and proto testTime is zero`() {
        val date = LocalDate.of(2026, 4, 8)
        val proto = PbSpo2TestResult.newBuilder()
            .setRecordingDevice("0004BF3D")
            .setTimeZoneOffset(180)
            .setTestStatus(PbSpo2TestStatus.SPO2_TEST_PASSED)
            .buildPartial()

        val result = PolarTestUtils.mapSpo2TestProto(proto, date, "BADDIR")

        assertNull(result.testTime)
    }

    @Test
    fun `mapSpo2TestProto() maps all proto fields correctly`() {
        val date = LocalDate.of(2026, 4, 14)
        val proto = PbSpo2TestResult.newBuilder()
            .setRecordingDevice("0004BF3D")
            .setTimeZoneOffset(180)
            .setTestStatus(PbSpo2TestStatus.SPO2_TEST_PASSED)
            .setBloodOxygenPercent(95)
            .setSpo2Class(PbSpo2Class.SPO2_CLASS_NORMAL)
            .setSpo2ValueDeviationFromBaseline(PbDeviationFromBaseline.DEVIATION_NO_BASELINE)
            .setSpo2QualityAveragePercent(99.0f)
            .setAverageHeartRateBpm(66)
            .setHeartRateVariabilityMs(79.97114f)
            .setSpo2HrvDeviationFromBaseline(PbDeviationFromBaseline.DEVIATION_USUAL)
            .setAltitudeMeters(18.13582f)
            .buildPartial()

        val result = PolarTestUtils.mapSpo2TestProto(proto, date, "063635")

        assertEquals("0004BF3D", result.recordingDevice)
        assertEquals(180, result.timeZoneOffsetMinutes)
        assertEquals(Spo2TestStatus.PASSED, result.testStatus)
        assertEquals(95, result.bloodOxygenPercent)
        assertEquals(Spo2Class.NORMAL, result.spo2Class)
        assertEquals(DeviationFromBaseline.NO_BASELINE, result.spo2ValueDeviationFromBaseline)
        assertEquals(99.0f, result.spo2QualityAveragePercent)
        assertEquals(66u, result.averageHeartRateBpm)
        assertEquals(79.97114f, result.heartRateVariabilityMs)
        assertEquals(DeviationFromBaseline.USUAL, result.spo2HrvDeviationFromBaseline)
        assertEquals(18.13582f, result.altitudeMeters)
        assertEquals("2026-04-14 06:36:35", result.testTime)
    }

    @Test
    fun `mapSpo2TestProto() maps ABOVE_USUAL hrv deviation`() {
        val date = LocalDate.of(2026, 4, 14)
        val proto = PbSpo2TestResult.newBuilder()
            .setTimeZoneOffset(180)
            .setTestStatus(PbSpo2TestStatus.SPO2_TEST_PASSED)
            .setSpo2HrvDeviationFromBaseline(PbDeviationFromBaseline.DEVIATION_ABOVE_USUAL)
            .setBloodOxygenPercent(96)
            .buildPartial()

        val result = PolarTestUtils.mapSpo2TestProto(proto, date, "063751")

        assertEquals(DeviationFromBaseline.ABOVE_USUAL, result.spo2HrvDeviationFromBaseline)
        assertEquals(96, result.bloodOxygenPercent)
    }

    @Test
    fun `mapSpo2TestProto() returns null for optional fields when not set in proto`() {
        val date = LocalDate.of(2026, 4, 8)
        val proto = PbSpo2TestResult.newBuilder()
            .setTimeZoneOffset(180)
            .setTestStatus(PbSpo2TestStatus.SPO2_TEST_PASSED)
            .buildPartial()

        val result = PolarTestUtils.mapSpo2TestProto(proto, date, "083906")

        assertNull(result.bloodOxygenPercent)
        assertNull(result.spo2Class)
        assertNull(result.spo2ValueDeviationFromBaseline)
        assertNull(result.spo2QualityAveragePercent)
        assertNull(result.averageHeartRateBpm)
        assertNull(result.heartRateVariabilityMs)
        assertNull(result.spo2HrvDeviationFromBaseline)
        assertNull(result.altitudeMeters)
    }

    @Test
    fun `mapSpo2TestEntry() parses proto bytes and maps to PolarSpo2TestData`() {
        val date = LocalDate.of(2026, 4, 8)
        val protoBytes = ByteArrayOutputStream().apply {
            PbSpo2TestResult.newBuilder()
                .setRecordingDevice("0004BF3D")
                .setTimeZoneOffset(180)
                .setTestStatus(PbSpo2TestStatus.SPO2_TEST_PASSED)
                .setBloodOxygenPercent(97)
                .setAverageHeartRateBpm(63)
                .buildPartial()
                .writeTo(this)
        }.toByteArray()

        val entry = Spo2TestEntry(date = date, timeDirName = "083906", protoBytes = protoBytes)
        val result = PolarTestUtils.mapSpo2TestEntry(entry)

        assertEquals("0004BF3D", result.recordingDevice)
        assertEquals(97, result.bloodOxygenPercent)
        assertEquals(63u, result.averageHeartRateBpm)
        assertEquals("2026-04-08 08:39:06", result.testTime)
    }

    // -----------------------------------------------------------------------------------------
    // Helpers
    // -----------------------------------------------------------------------------------------

    private fun requestFor(path: String): ByteArray =
        PftpRequest.PbPFtpOperation.newBuilder()
            .setCommand(PftpRequest.PbPFtpOperation.Command.GET)
            .setPath(path)
            .build()
            .toByteArray()

    private fun buildDirectory(vararg entryNames: String): ByteArrayOutputStream =
        ByteArrayOutputStream().apply {
            val entries = entryNames.map {
                PbPFtpEntry.newBuilder().setName(it).setSize(0L).buildPartial()
            }
            write(PbPFtpDirectory.newBuilder().addAllEntries(entries).buildPartial().toByteArray())
        }

    private fun buildProto(block: PbSpo2TestResult.Builder.() -> PbSpo2TestResult.Builder): ByteArrayOutputStream =
        ByteArrayOutputStream().apply {
            PbSpo2TestResult.newBuilder().block().buildPartial().writeTo(this)
        }
}

File: sources/iOS/ios-communications/Tests/PolarBleSdkTests/PolarTestUtilsTest.swift
//  Copyright © 2026 Polar. All rights reserved.

import XCTest
import RxSwift
import RxTest
@testable import PolarBleSdk

class PolarTestUtilsTests: XCTestCase {

    var mockClient: MockBlePsFtpClient!

    private let dateFormatter: DateFormatter = {
        let f = DateFormatter()
        f.dateFormat = "yyyyMMdd"
        f.locale = Locale(identifier: "en_US_POSIX")
        return f
    }()

    override func setUpWithError() throws {
        mockClient = MockBlePsFtpClient(gattServiceTransmitter: MockGattServiceTransmitterImpl())
    }

    override func tearDownWithError() throws {
        mockClient = nil
    }

    // MARK: - Success: full proto

    func testReadSpo2TestFromDayDirectory_Success_AllFields() throws {
        // Arrange
        let date = makeDate(year: 2026, month: 4, day: 13)
        let proto = buildSpo2TestProto()
        mockClient.requestReturnValues = try makeDirThenFileSingles(proto: proto)

        // Act
        let result = try awaitFirst(
            PolarTestUtils.readSpo2TestFromDayDirectory(client: mockClient, date: date))

        // Assert – all non-zero/non-empty fields should be mapped
        XCTAssertEqual(result.bloodOxygenPercent, 97)
        XCTAssertEqual(result.testStatus, .passed)
        XCTAssertEqual(result.spo2Class, .normal)
        XCTAssertEqual(result.averageHeartRateBpm, 65)
        XCTAssertEqual(result.recordingDevice, "Polar Vantage V3")
        XCTAssertEqual(result.spo2ValueDeviationFromBaseline, .usual)
        XCTAssertEqual(result.spo2HrvDeviationFromBaseline, .noBaseline)
        XCTAssertNil(result.triggerType)  // triggerType is not in the proto
        XCTAssertEqual(result.timeZoneOffsetMinutes, 120)
        XCTAssertEqual(result.spo2QualityAveragePercent ?? 0, Float(90.0), accuracy: 0.001)
        XCTAssertEqual(result.heartRateVariabilityMs ?? 0, 45.5, accuracy: 0.001)
        XCTAssertEqual(result.altitudeMeters ?? 0, 100.0, accuracy: 0.001)
    }

    // MARK: - Success: date from testTime proto field (UInt64 ms since epoch)

    func testReadSpo2TestFromDayDirectory_Success_UsesTestTimeWhenPresent() throws {
        // Arrange – testTime = ms since epoch for 2026-04-13 14:25:07 UTC
        // 1744554307000 ms = 2026-04-13T14:25:07Z
        let fallbackDate = makeDate(year: 2026, month: 4, day: 13)
        var proto = Data_PbSpo2TestResult()
        proto.bloodOxygenPercent = 95
        proto.testTime = 1744554307000  // 2026-04-13 14:25:07 UTC
        mockClient.requestReturnValues = try makeDirThenFileSingles(proto: proto)

        // Act
        let result = try awaitFirst(
            PolarTestUtils.readSpo2TestFromDayDirectory(client: mockClient, date: fallbackDate)
        )

        // Assert – result date should not equal fallback (testTime was decoded)
        XCTAssertNotEqual(result.date, fallbackDate)
        // The date should be in the right ballpark (2026-04-13)
        let components = Calendar(identifier: .gregorian).dateComponents([.year, .month, .day], from: result.date)
        XCTAssertEqual(components.year, 2026)
        XCTAssertEqual(components.month, 4)
        XCTAssertEqual(components.day, 13)
    }

    // MARK: - Success: fallback date used when testTime is zero

    func testReadSpo2TestFromDayDirectory_Success_FallsBackToDateWhenTestTimeAbsent() throws {
        // Arrange – testTime left at default 0
        let fallbackDate = makeDate(year: 2026, month: 4, day: 13)
        var proto = Data_PbSpo2TestResult()
        proto.bloodOxygenPercent = 95
        // testTime == 0 → treated as absent
        mockClient.requestReturnValues = try makeDirThenFileSingles(proto: proto)

        // Act
        let result = try awaitFirst(
            PolarTestUtils.readSpo2TestFromDayDirectory(client: mockClient, date: fallbackDate)
        )

        // Assert – when folder name is "142507" (from makeDirThenFileSingles default), date should be parsed from folder
        // but since we used fallbackDate as dayDate the year/month/day should match
        let components = Calendar(identifier: .gregorian).dateComponents([.year, .month, .day], from: result.date)
        XCTAssertEqual(components.year, 2026)
        XCTAssertEqual(components.month, 4)
        XCTAssertEqual(components.day, 13)
    }

    // MARK: - triggerType is always nil (not in proto)

    func testReadSpo2TestFromDayDirectory_Success_ZeroValuesMapToNil() throws {
        // Arrange – empty proto: all numeric fields are zero, strings empty
        let proto = Data_PbSpo2TestResult()
        mockClient.requestReturnValues = try makeDirThenFileSingles(proto: proto)

        // Act
        let result = try awaitFirst(
            PolarTestUtils.readSpo2TestFromDayDirectory(client: mockClient, date: Date())
        )

        // Assert – optional fields should be nil when proto value is zero / empty
        XCTAssertNil(result.recordingDevice)
        XCTAssertNil(result.timeZoneOffsetMinutes)
        XCTAssertNil(result.bloodOxygenPercent)
        XCTAssertNil(result.spo2QualityAveragePercent)
        XCTAssertNil(result.averageHeartRateBpm)
        XCTAssertNil(result.heartRateVariabilityMs)
        XCTAssertNil(result.altitudeMeters)
    }

    // MARK: - Success: correct paths are requested

    func testReadSpo2TestFromDayDirectory_CorrectPathsRequested() throws {
        // Arrange
        let date = makeDate(year: 2026, month: 4, day: 13)
        let proto = buildSpo2TestProto()
        mockClient.requestReturnValues = try makeDirThenFileSingles(proto: proto, subDirName: "142507")

        _ = try awaitAll(
            PolarTestUtils.readSpo2TestFromDayDirectory(client: mockClient, date: date)
        )

        // Assert – first call lists the SPO2TEST directory, second fetches the file
        XCTAssertEqual(mockClient.requestCalls.count, 2)

        let firstOp = try Protocol_PbPFtpOperation(serializedBytes: mockClient.requestCalls[0])
        XCTAssertEqual(firstOp.path, "/U/0/20260413/SPO2TEST/")

        let secondOp = try Protocol_PbPFtpOperation(serializedBytes: mockClient.requestCalls[1])
        XCTAssertEqual(secondOp.path, "/U/0/20260413/SPO2TEST/142507/SPO2TRES.BPB")
    }

    // MARK: - Empty directory (no subdirectory entries) → completes empty

    func testReadSpo2TestFromDayDirectory_EmptyDirectory_CompletesEmpty() throws {
        // Arrange – SPO2TEST/ directory exists but has no entries
        var emptyDir = Protocol_PbPFtpDirectory()
        emptyDir.entries = []
        mockClient.requestReturnValues = [Single.just(try emptyDir.serializedData())]

        // Act & Assert
        let results = try awaitAll(
            PolarTestUtils.readSpo2TestFromDayDirectory(client: mockClient, date: Date())
        )
        XCTAssertTrue(results.isEmpty)
    }

    // MARK: - Directory has only file entries (no subdirectory) → completes empty

    func testReadSpo2TestFromDayDirectory_DirectoryHasNoSubdirs_CompletesEmpty() throws {
        // Arrange – entry without trailing slash is a file, not a subdir
        var dir = Protocol_PbPFtpDirectory()
        var fileEntry = Protocol_PbPFtpEntry()
        fileEntry.name = "SPO2TRES.BPB"   // no trailing slash
        fileEntry.size = 0
        dir.entries = [fileEntry]
        mockClient.requestReturnValues = [Single.just(try dir.serializedData())]

        // Act & Assert
        let results = try awaitAll(
            PolarTestUtils.readSpo2TestFromDayDirectory(client: mockClient, date: Date())
        )
        XCTAssertTrue(results.isEmpty)
    }

    // MARK: - Directory listing fails → completes empty

    func testReadSpo2TestFromDayDirectory_DirectoryListingFails_CompletesEmpty() {
        // Arrange – simulate device error 103 (file/directory not found)
        mockClient.requestReturnValues = [
            Single.error(NSError(domain: "BLE", code: 103, userInfo: nil))
        ]

        let results = try! awaitAll(
            PolarTestUtils.readSpo2TestFromDayDirectory(client: mockClient, date: Date())
        )
        XCTAssertTrue(results.isEmpty)
    }

    // MARK: - File fetch fails after directory found → completes empty

    func testReadSpo2TestFromDayDirectory_FileFetchFails_CompletesEmpty() throws {
        // Arrange – directory listing succeeds, but file request errors
        var dir = Protocol_PbPFtpDirectory()
        var entry = Protocol_PbPFtpEntry()
        entry.name = "142507/"
        entry.size = 0
        dir.entries = [entry]

        mockClient.requestReturnValues = [
            Single.just(try dir.serializedData()),
            Single.error(NSError(domain: "BLE", code: 103, userInfo: nil))
        ]

        // Act & Assert
        let results = try awaitAll(
            PolarTestUtils.readSpo2TestFromDayDirectory(client: mockClient, date: Date())
        )
        XCTAssertTrue(results.isEmpty)
    }

    // MARK: - Malformed file data → completes empty

    func testReadSpo2TestFromDayDirectory_MalformedFileData_CompletesEmpty() throws {
        // Arrange – directory listing succeeds but the proto bytes are garbage
        var dir = Protocol_PbPFtpDirectory()
        var entry = Protocol_PbPFtpEntry()
        entry.name = "142507/"
        entry.size = 0
        dir.entries = [entry]

        mockClient.requestReturnValues = [
            Single.just(try dir.serializedData()),
            Single.just(Data([0xFF, 0xFE, 0xFD]))
        ]

        // Act & Assert
        let results = try awaitAll(
            PolarTestUtils.readSpo2TestFromDayDirectory(client: mockClient, date: Date())
        )
        XCTAssertTrue(results.isEmpty)
    }

    // MARK: - Multiple subdirectories → all are returned

    func testReadSpo2TestFromDayDirectory_MultipleSubdirs_ReturnsAll() throws {
        // Arrange – two time subdirectories; both should be read and returned
        var dir = Protocol_PbPFtpDirectory()
        var entry1 = Protocol_PbPFtpEntry(); entry1.name = "093635/"; entry1.size = 0
        var entry2 = Protocol_PbPFtpEntry(); entry2.name = "093751/"; entry2.size = 0
        dir.entries = [entry1, entry2]

        var proto1 = buildSpo2TestProto()
        proto1.bloodOxygenPercent = 97

        var proto2 = buildSpo2TestProto()
        proto2.bloodOxygenPercent = 95

        mockClient.requestReturnValues = [
            Single.just(try dir.serializedData()),
            Single.just(try proto1.serializedData()),
            Single.just(try proto2.serializedData())
        ]

        let date = makeDate(year: 2026, month: 4, day: 13)
        let results = try awaitAll(
            PolarTestUtils.readSpo2TestFromDayDirectory(client: mockClient, date: date)
        )

        // Both subdirectories should produce a result
        XCTAssertEqual(results.count, 2)
        let oxygenValues = Set(results.compactMap { $0.bloodOxygenPercent })
        XCTAssertEqual(oxygenValues, Set([97, 95]))

        // Both file paths should have been requested
        XCTAssertEqual(mockClient.requestCalls.count, 3)
        let paths = try mockClient.requestCalls.dropFirst().map {
            try Protocol_PbPFtpOperation(serializedBytes: $0).path
        }
        XCTAssertTrue(paths.contains("/U/0/20260413/SPO2TEST/093635/SPO2TRES.BPB"))
        XCTAssertTrue(paths.contains("/U/0/20260413/SPO2TEST/093751/SPO2TRES.BPB"))
    }

    // MARK: - Enum mapping: Spo2Class

    func testSpo2ClassMapping() {
        XCTAssertEqual(PolarSpo2TestData.Spo2Class(rawValue: 0), .unknown)
        XCTAssertEqual(PolarSpo2TestData.Spo2Class(rawValue: 1), .veryLow)
        XCTAssertEqual(PolarSpo2TestData.Spo2Class(rawValue: 2), .low)
        XCTAssertEqual(PolarSpo2TestData.Spo2Class(rawValue: 3), .normal)
        XCTAssertNil(PolarSpo2TestData.Spo2Class(rawValue: 99))
    }

    // MARK: - Enum mapping: Spo2TestStatus

    func testSpo2TestStatusMapping() {
        XCTAssertEqual(PolarSpo2TestData.Spo2TestStatus(rawValue: 0), .passed)
        XCTAssertEqual(PolarSpo2TestData.Spo2TestStatus(rawValue: 1), .inconclusiveTooLowQualityInSamples)
        XCTAssertEqual(PolarSpo2TestData.Spo2TestStatus(rawValue: 2), .inconclusiveTooLowOverallQuality)
        XCTAssertEqual(PolarSpo2TestData.Spo2TestStatus(rawValue: 3), .inconclusiveTooManyMissingSamples)
        XCTAssertNil(PolarSpo2TestData.Spo2TestStatus(rawValue: 99))
    }

    // MARK: - Enum mapping: DeviationFromBaseline

    func testDeviationFromBaselineMapping() {
        XCTAssertEqual(PolarSpo2TestData.DeviationFromBaseline(rawValue: 0), .noBaseline)
        XCTAssertEqual(PolarSpo2TestData.DeviationFromBaseline(rawValue: 1), .belowUsual)
        XCTAssertEqual(PolarSpo2TestData.DeviationFromBaseline(rawValue: 2), .usual)
        XCTAssertEqual(PolarSpo2TestData.DeviationFromBaseline(rawValue: 3), .aboveUsual)
        XCTAssertNil(PolarSpo2TestData.DeviationFromBaseline(rawValue: 99))
    }

    // MARK: - Enum mapping: Spo2TestTriggerType

    func testTriggerTypeMapping() {
        XCTAssertEqual(PolarSpo2TestData.Spo2TestTriggerType(rawValue: 0), .manual)
        XCTAssertEqual(PolarSpo2TestData.Spo2TestTriggerType(rawValue: 1), .automatic)
        XCTAssertNil(PolarSpo2TestData.Spo2TestTriggerType(rawValue: 99))
    }

    // MARK: - Enum mapping: each Spo2TestStatus value round-trips from proto

    func testAllSpo2TestStatuses_RoundTripFromProto() throws {
        let statuses: [(Data_PbSpo2TestStatus, PolarSpo2TestData.Spo2TestStatus)] = [
            (.spo2TestPassed, .passed),
            (.spo2TestInconclusiveTooLowQualityInSamples, .inconclusiveTooLowQualityInSamples),
            (.spo2TestInconclusiveTooLowOverallQuality, .inconclusiveTooLowOverallQuality),
            (.spo2TestInconclusiveTooManyMissingSamples, .inconclusiveTooManyMissingSamples)
        ]
        for (protoStatus, expected) in statuses {
            var proto = Data_PbSpo2TestResult()
            proto.testStatus = protoStatus
            proto.bloodOxygenPercent = 95
            mockClient = MockBlePsFtpClient(gattServiceTransmitter: MockGattServiceTransmitterImpl())
            mockClient.requestReturnValues = try makeDirThenFileSingles(proto: proto)

            let result = try awaitFirst(
                PolarTestUtils.readSpo2TestFromDayDirectory(client: mockClient, date: Date())
            )
            XCTAssertEqual(result.testStatus, expected, "testStatus mismatch for \(protoStatus)")
        }
    }

    // MARK: - Enum mapping: each Spo2Class value round-trips from proto

    func testAllSpo2Classes_RoundTripFromProto() throws {
        let classes: [(Data_PbSpo2Class, PolarSpo2TestData.Spo2Class)] = [
            (.spo2ClassUnknown, .unknown),
            (.spo2ClassVeryLow, .veryLow),
            (.spo2ClassLow, .low),
            (.spo2ClassNormal, .normal)
        ]
        for (protoClass, expected) in classes {
            var proto = Data_PbSpo2TestResult()
            proto.spo2Class = protoClass
            proto.bloodOxygenPercent = 95
            mockClient = MockBlePsFtpClient(gattServiceTransmitter: MockGattServiceTransmitterImpl())
            mockClient.requestReturnValues = try makeDirThenFileSingles(proto: proto)

            let result = try awaitFirst(
                PolarTestUtils.readSpo2TestFromDayDirectory(client: mockClient, date: Date())
            )
            XCTAssertEqual(result.spo2Class, expected, "spo2Class mismatch for \(protoClass)")
        }
    }

    // MARK: - Helpers

    /// Build a fully-populated SPO2 test proto using the real generated field names.
    private func buildSpo2TestProto() -> Data_PbSpo2TestResult {
        var proto = Data_PbSpo2TestResult()
        proto.recordingDevice = "Polar Vantage V3"
        proto.testStatus = .spo2TestPassed
        proto.bloodOxygenPercent = 97
        proto.spo2Class = .spo2ClassNormal
        proto.spo2ValueDeviationFromBaseline = .deviationUsual
        proto.spo2HrvDeviationFromBaseline = .deviationNoBaseline
        proto.spo2QualityAveragePercent = 90.0
        proto.averageHeartRateBpm = 65
        proto.heartRateVariabilityMs = 45.5
        proto.altitudeMeters = 100.0
        proto.timeZoneOffset = 120
        proto.testTime = 1744574400000  // 2026-04-14 00:00:00 UTC
        return proto
    }

    /// Build [directorySingle, fileSingle] to feed requestReturnValues.
    private func makeDirThenFileSingles(
        proto: Data_PbSpo2TestResult,
        subDirName: String = "142507"
    ) throws -> [Single<Data>] {
        var dir = Protocol_PbPFtpDirectory()
        var entry = Protocol_PbPFtpEntry()
        entry.name = "\(subDirName)/"
        entry.size = 0
        dir.entries = [entry]
        return [
            Single.just(try dir.serializedData()),
            Single.just(try proto.serializedData())
        ]
    }

    /// Convenience: collect all emitted values from an Observable.
    @discardableResult
    private func awaitAll<T>(_ observable: Observable<T>, timeout: TimeInterval = 2.0) throws -> [T] {
        var results: [T] = []
        var capturedError: Error?
        let exp = expectation(description: "awaitAll")
        let disposable = observable.subscribe(
            onNext: { results.append($0) },
            onError: { capturedError = $0; exp.fulfill() },
            onCompleted: { exp.fulfill() }
        )
        wait(for: [exp], timeout: timeout)
        disposable.dispose()
        if let error = capturedError { throw error }
        return results
    }

    /// Convenience: assert the Observable emits at least one value and return the first.
    @discardableResult
    private func awaitFirst<T>(_ observable: Observable<T>, timeout: TimeInterval = 2.0) throws -> T {
        let results = try awaitAll(observable, timeout: timeout)
        return try XCTUnwrap(results.first, "Expected at least one emitted value")
    }

    /// Create a Date at midnight UTC for a given year/month/day.
    private func makeDate(year: Int, month: Int, day: Int) -> Date {
        var c = DateComponents()
        c.year = year; c.month = month; c.day = day
        c.hour = 0; c.minute = 0; c.second = 0
        return Calendar(identifier: .gregorian).date(from: c)!
    }
}

